File: naru_agent/skills/registry.py
from __future__ import annotations

import logging
from concurrent.futures import ThreadPoolExecutor, as_completed

from naru_agent.skills.base import BaseSkill, SkillContext, SkillResult
from naru_agent.skills.selectors import BaseSkillSelector, KeywordSkillSelector

logger = logging.getLogger(__name__)


class SkillRegistry:
    """Registry that selects and runs skills in parallel."""

    def __init__(
        self,
        skills: list[BaseSkill],
        selector: BaseSkillSelector | None = None,
        max_active_skills: int = 3,
    ):
        self.skills = skills
        self.selector = selector or KeywordSkillSelector()
        self.max_active_skills = max_active_skills

    def select_skills(self, message: str) -> list[BaseSkill]:
        selected = self.selector.select(self.skills, message)
        # Sort by priority descending, take top max_active_skills
        selected.sort(key=lambda s: s.priority, reverse=True)
        return selected[: self.max_active_skills]

    def run_skills(self, message: str, context: SkillContext) -> list[SkillResult]:
        selected = self.select_skills(message)
        if not selected:
            return []

        results: list[SkillResult] = []

        if len(selected) == 1:
            # Skip thread pool overhead for the common single-skill case
            try:
                r = selected[0].run(message, context)
                results.append(r)
            except Exception:
                logger.warning("Skill '%s' failed", selected[0].name, exc_info=True)
        else:
            with ThreadPoolExecutor(max_workers=len(selected)) as pool:
                future_to_skill = {
                    pool.submit(s.run, message, context): s for s in selected
                }
                for future in as_completed(future_to_skill):
                    s = future_to_skill[future]
                    try:
                        r = future.result()
                        results.append(r)
                    except Exception:
                        logger.warning("Skill '%s' failed", s.name, exc_info=True)

        # Sort by priority descending
        skill_priority = {s.name: s.priority for s in selected}
        results.sort(key=lambda r: skill_priority.get(r.skill_name, 0), reverse=True)
        return results

File: js/src/skills/registry.ts
import type { BaseSkill, SkillContext, SkillResult } from "./base.js";
import type { BaseSkillSelector } from "./selectors.js";
import { KeywordSkillSelector } from "./selectors.js";
import { makeSkillResult } from "./base.js";

export class SkillRegistry {
  private skills: BaseSkill[];
  private selector: BaseSkillSelector;
  private maxActiveSkills: number;

  constructor(
    skills: BaseSkill[],
    selector?: BaseSkillSelector,
    maxActiveSkills = 3,
  ) {
    this.skills = skills;
    this.selector = selector ?? new KeywordSkillSelector();
    this.maxActiveSkills = maxActiveSkills;
  }

  async selectSkills(message: string): Promise<BaseSkill[]> {
    const selected = await this.selector.select(this.skills, message);
    return selected
      .sort((a, b) => b.priority - a.priority)
      .slice(0, this.maxActiveSkills);
  }

  async runSkills(
    message: string,
    context: SkillContext,
  ): Promise<SkillResult[]> {
    const selected = await this.selectSkills(message);
    if (selected.length === 0) return [];

    const results = await Promise.allSettled(
      selected.map(async (s) => {
        try {
          const result = await s.run(message, context);
          return { ...result, skillName: s.name };
        } catch {
          return makeSkillResult({ skillName: s.name, skipped: true });
        }
      }),
    );

    return results
      .filter(
        (r): r is PromiseFulfilledResult<SkillResult> =>
          r.status === "fulfilled",
      )
      .map((r) => r.value)
      .filter((r) => !r.skipped);
  }
}
